cityofaustin/atd-moped
# Edit existing component selected features and attributes

## Associated issues
https://github.com/cityofaustin/atd-data-tech/issues/10626

This PR updates the component map to enable update of line segment and point features selected from CTN layers. **This does not includes drawn components feature edits - PR to come for that.**

⚠️ This is a big diff so I left many comments throughout. Lots of the big chunks are existing code reorganized so I noted that when I could. Please ping me if talking through some of the code would help with review. If I could do it again, I would have split the attributes and feature editing into two PRs.

**Summary of changes**
- Split `ComponentsSourcesAndLayers` for the map into:
   - `DraftComponentSourcesAndLayers` (adding features to a new component)
   - `ProjectSourcesAndLayers` (existing component features)
   - `ClickedComponentSourcesAndLayers` (features highlighted by clicking a component list item)
   - `EditDraftComponentSourcesAndLayers` (editing an existing component's features)
   - `CTNSourcesAndLayers` (shows selectable/clickable line segments and points when creating or editing)
- Added `useCreateComponent`, `useUpdateComponent`, `useDeleteComponent` hooks to organize state and handlers related to component CRUD
- Refactored create and update state to use `useReducer` to organize map action state changes and also reduce the amount state passed around 
- Added `is_deleted` column to feature tables for soft deletes
- Added a unique constraint to project subcomponents (the chips in the attribtue form) on `project_component_id` and `subcomponent_id` so that we can flip `is_deleted` on conflict
- Filtering out soft deleted features in `uniform_features` view (not sure this is the best way to filter soft deletes but this feeds the `project_geography` view so that is why I chose it)

## Testing
**URL to test:** 

Local or https://md-10626-editing-components--atd-moped-main.netlify.app/moped/project

**Steps to test:**
1. Go to an existing or new project's **Map** tab
2. Add a new line component and choose:
    - Type: Bike Lane - Buffered
    - Subcomponents: Raised pavement markers
    - Description: any description but I like "this is a test"
3. Click **Save** on the form and then add some line segments from the dashed line CTN layer and then click **Save** in the left list item
4. Click the left list item of the component you just created and then clicked **Edit**
5. Choose **Attributes** and use the form to remove the subcomponent and update the description
6. Edit the attributes again and you should see the latest values from the edits in the last step
7. Now, choose **Map** when editing the component
8. You should see the existing line features shown in orange
9. Click an existing orange segment and it should disappear
10. Click a dashed segment to add a feature to the component
11. Click **Save Edit** and you should see your new features saved to the component
12. You can repeat these steps with a point component like one with type **Bike Box**

---
#### Ship list
